src/cpp/StatisticsBackendData.cpp

```diff
@@ -552,48 +541,134 @@ database::DatabaseEntityQueue* StatisticsBackendData::get_entity_queue()
+            database_->check_alerts_matching_entities();
         }
     }
 }
 
 void StatisticsBackendData::start_alert_watcher()
 {
-    std::unique_lock<std::mutex> lock(alert_watcher_mutex_);
     stop_alert_watcher_ = false;
     alert_watcher_thread_ = std::thread(&StatisticsBackendData::alert_watcher, this);
 }
 
 void StatisticsBackendData::stop_alert_watcher()
 {
+    stop_alert_watcher_ = true;
+    if (alert_watcher_thread_.joinable())
     {
-        std::lock_guard<std::mutex> lock(alert_watcher_mutex_);
-        stop_alert_watcher_ = true;
+        alert_watcher_thread_.join();
     }
+}
 
-    alert_watcher_cv_.notify_all();
+void StatisticsBackendData::start_topic_spy(
+        EntityId monitor_id,
+        const std::string& topic_name,
+        std::function<void(const std::string& data)> on_data_received)
+{
+    std::lock_guard<StatisticsBackendData> guard(*this);
 
-    if (alert_watcher_thread_.joinable())
+    auto monitor_it = monitors_by_entity_.find(monitor_id);
+    if (monitor_it == monitors_by_entity_.end())
     {
-        alert_watcher_thread_.join();
+        throw BadParameter("No monitor with such ID");
+    }
+    auto& monitor = monitor_it->second;
+
+    // If the topic is already being spied, do nothing
+    if (monitor->user_data_readers.find(topic_name) != monitor->user_data_readers.end())
+    {
+        EPROSIMA_LOG_WARNING(STATISTICS_BACKEND_DATA, "Topic '" << topic_name << "' is already being spied");
+        return;
+    }
+
+    fastdds::dds::Topic* topic = nullptr;
+
+    auto topic_it = monitor->user_data_topics.find(topic_name);
+    if (topic_it == monitor->user_data_topics.end())
+    {
+        fastdds::dds::DynamicType::_ref_type topic_type =
+                monitor->user_data_context.get_type_from_topic_name(topic_name);
+        if (!topic_type)
+        {
+            throw Error("Unable to get the type for topic '" + topic_name + "'");
+        }
+
+        std::string type_name = topic_type->get_name().to_string();
+        fastdds::dds::TypeSupport type_support =
+                fastdds::dds::TypeSupport(new fastdds::dds::DynamicPubSubType(topic_type));
+        if (fastdds::dds::RETCODE_OK != type_support.register_type(monitor->participant, type_name))
+        {
+            throw Error("Error registering type for topic '" + topic_name + "'");
+        }
+
+        fastdds::dds::TopicQos topic_qos;
+        topic = monitor->participant->create_topic(topic_name, type_name, topic_qos);
+        if (!topic)
+        {
+            throw Error("Error creating topic '" + topic_name + "'");
+        }
+        monitor->user_data_topics[topic_name] = topic;
+    }
+    else
+    {
+        topic = topic_it->second;
+    }
+
+    // Create a reader in the specified topic
+    fastdds::dds::DataReaderListener* listener = new subscriber::UserDataReaderListener(
+        on_data_received,
+        &monitor->user_data_context);
+    monitor->user_data_listeners[topic_name] = listener;
+
```

Comment:
In order to spy all partitions, the subscriber has to have the wildcard ("*") partition. The following code adds to the subscriptor the wildcard partition (without checking if this partition is already added).

```suggestion
fastdds::dds::SubscriberQos new_sub_qos = fastdds::dds::SUBSCRIBER_QOS_DEFAULT;
new_sub_qos.partition().push_back("*");
monitor->subscriber->set_qos(new_sub_qos);
```